### PR TITLE
[stable/traefik] Allow replica count to be specified

### PR DESCRIPTION
**Pull request by cknowles**
For https://github.com/kubernetes/charts/issues/335. I've avoided bumping the Chart version for now to avoid clashes with other PRs.

cc @krancour the chart maintainer

### CODE REVIEW

**Comment by krancour** [comment]
There's more involved in HA Traefik than just increasing the replica count. I expect that this approach will work fine for many cases, but will rapidly lead to trouble for anyone getting certs from Let's Encrypt. Without arranging Traefik nodes properly into a master/slave configuration, _each_ node will end up requesting certs and that's going to cut quickly into the very modest rate limit that Let's Encrypt enforces. If we accept this PR, I think the doc needs to very clearly state that this is a naive approach to scaling Traefik and should specifically call out the scenario I just articulated as a reason that the mileage one gets from turning this nob may vary.

**Comment by cknowles** [comment]
Make sense. I did this purely to address upgrade/downtime but it doesn't accommodate election at all. Since this is still useful for many use cases, I think it would be good to merge it but also work on the election additions as soon as possible.

**Comment by cknowles** [comment]
@krancour I think it's clear enough, what do you think?

File: stable/traefik/README.md
Comment (krancour) [inline stable/traefik/README.md]: Typo here. s/with/which/. If additionally you don't mind some formatting suggestions:

```
The number of replicas to run; __NOTE:__ Full Traefik clustering with leader election is not yet supported, which can affect any configured Let's Encrypt setup; see Clustering section
```

File: stable/traefik/README.md
Comment (krancour) [inline stable/traefik/README.md]: Suggested edit:

```
Currently, it is possible to specify the number of `replicas`, but the implementation is naive.
```

File: stable/traefik/README.md
Comment (krancour) [inline stable/traefik/README.md]: Can we add blank lines above and below this (so it reads more like a heading)?

File: stable/traefik/README.md
Comment (krancour) [inline stable/traefik/README.md]: s/i.e.//

Preference: s/master/keader/... I think it reads better that way.

File: stable/traefik/README.md
Comment (krancour) [inline stable/traefik/README.md]: Can we merge this paragraph with the one above, since they're both brief and both are supporting the same general bit of advice? Might need to re-write a little after combining them.

**Review by krancour — CHANGES_REQUESTED** [review]

File: stable/traefik/templates/deployment.yaml
Comment (krancour) [inline stable/traefik/templates/deployment.yaml]: Can we set a default right here in case `replicas` isn't set at all? (i.e. in case someone continues using their _old_, customized `values.yaml` with the updated chart)? Something like this:

```
replicas: {{ .Values.replicas | default 1 }}
```

**Review by krancour — CHANGES_REQUESTED** [review]

File: stable/traefik/templates/deployment.yaml
Comment (cknowles) [reply stable/traefik/templates/deployment.yaml]: Sure, I only realised this after. Due to the way user settings are merged in with the default `values.yaml`, I think generally defaults have to live in the templates anyway.

**Comment by cknowles** [comment]
@krancour done those bits, see what you think.

**Comment by krancour** [comment]
@c-knowles I like the caveats you added. I'm having some renewed concern over this naive approach to scaling after a similar discussion on #477. I understand your use case, but I have a sincere worry that every new option we add, no matter what level of caveats come with it, is an option that some people will inevitably be tempted into exercising. I'm starting to think that in such cases, it might be better for people to consume such features from a fork of this chart.

This isn't the final word on it. I'd really like to hear the perspectives of other maintainers, @lachie83 especially.

**Comment by cknowles** [comment]
@krancour Sure, I understand. FYI I've already launch this to our integration environment but not planning on launching it to production soon. So far it's been working fine with Let's Encrypt turned off and provides some resiliency. I have noticed the dashboard UI seems to disconnect more often than I'd hoped, not sure if that's a more general issue as we just starting using Traefik.

**Comment by lachie83** [comment]
@krancour @c-knowles I think that given that the default configuration is with LetsEncrypt disabled scaling the replicas is okay. Is this the only thing that breaks if you scale? If that's the case then I'm okay with making it configurable.

**Comment by cknowles** [comment]
@lachie83 Yes, I believe that is the only thing that breaks with scaling. I've been running this without issues in our integration environment for several days so far. The dashboard issue I mentioned above I managed to nail down to an instability in the k8s cluster itself (docker hanging).

**Comment by krancour** [comment]
@lachie83 fair point made that this only (as far as we know) affects use of Let's Encrypt, which the chart does not enable by default anyway.

One thing I'd like to get a handle on is if or how the decision to move forward with this affects our disposition toward #477 (running Traefik as a daemonset). I still perceive that as a bad idea, but in some ways, it's not dissimilar from this PR.

**Comment by lachie83** [comment]
@c-knowles @krancour Can we come to a consensus as to whether or not to merge this?

**Comment by krancour** [comment]
I vote yes. Let's just make sure the version string is correct... then 👍

**Comment by lachie83** [comment]
@k8s-bot ok to test

**Pull request merged**
